### PR TITLE
BGP Polling: Add error code management

### PR DESCRIPTION
**Pull request by kedare**
DO NOT DELETE THIS TEXT

Poll BGP errors codes from devices (Tested on IOS XR, JunOS, Arista EOS)
Display them on the routing pages

First line of status is generated from the error codes and subcodes (if supported)
The second line is that the device return as error text (if supported)

Screenshot 2020-04-15 at 11 47 35

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by kedare** [comment]
I am adding the missing test data

**Comment by kedare** [comment]
I am still adding test data, I suppose I can just add blank data like most of the other values ? (new fields to Null) as I can't add real test data (as it looks like the cisco ios is inherited by a lot of others OS I don't have access to)

**Comment by kedare** [comment]
Tests passed, I can't do much for the code climate issues however ?

File: database/migrations/2020_04_13_150500_add_last_error_fields_to_bgp_peers.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+     *
+     *
+     *
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::table('bgpPeers', function (Blueprint $table) {
+            $table->dropColumn('bgpPeerLastErrorCode');
```
Comment (murrant) [inline database/migrations/2020_04_13_150500_add_last_error_fields_to_bgp_peers.php:34]: This should be one dropColumn() with an array of all columns to drop.

File: includes/functions.php [hunk 2]
```diff
@@ -2524,3 +2524,94 @@ function oxidized_node_update($hostname, $msg, $username = 'not_provided')
...
+/**
+ * @params int code
+ * @params int subcode
+ * @return string
+ * Take a BGP error code and subcode to return a string representation of it
+ */
+function describe_bgp_error_code($code, $subcode)
+{
+    // https://www.iana.org/assignments/bgp-parameters/bgp-parameters.xhtml#bgp-parameters-3
+    $codes = [
```
Comment (murrant) [inline includes/functions.php:2538]: The best way to do this is to create a new language file (resources/lang/en/bgp.php) and add all these messages there, that way you can look them up like `trans('bgp.error_codes.$code")`  It will also allow translation :)

File: includes/functions.php [hunk 3]
```diff
@@ -2524,3 +2524,94 @@ function oxidized_node_update($hostname, $msg, $username = 'not_provided')
...
+    }
+
+    if (array_key_exists($code, $subcodes) && array_key_exists($subcode, $subcodes[$code])) {
+        $subcode_message = $subcodes[$code][$subcode];
+    } else {
+        $subcode_message = "";
+    }
+
+    if ($subcode_message == "") {
+        return "[" . $code . " " . $subcode . "]: " . $code_message;
```
Comment (murrant) [inline includes/functions.php:2613]: Does including the code and subcode improve things?  I felt a little confused as to what the numbers were looking at the webui.

File: includes/polling/bgp-peers.inc.php [hunk 4]
```diff
@@ -243,7 +266,7 @@
                                 // if parsing fails, leave the data as-is
                             }
                         }
-
+
```
Comment (murrant) [inline includes/polling/bgp-peers.inc.php:269]: oopsie whitespace

**Comment by murrant** [comment]
Looks good, just some improvements.

It might be a good idea to include some updated test data with the new fields.  They seem to all be null.

File: includes/functions.php (hunk 3 above)
Comment (kedare) [reply includes/functions.php:2613]: Indeed we can remove them from the WebUI

**Comment by kedare** [comment]
For the test data I could potentially add some for Arista, JunOS, and IOS-XR, but I will have to add it by hand on the test data, it's not very clear how to do that on the snmp test data ?

File: includes/polling/bgp-peers.inc.php (hunk 4 above)
Comment (kedare) [reply includes/polling/bgp-peers.inc.php]: Fixed

File: includes/functions.php (hunk 3 above)
Comment (kedare) [reply includes/functions.php]: Done

File: includes/functions.php (hunk 2 above)
Comment (kedare) [reply includes/functions.php]: Done

File: database/migrations/2020_04_13_150500_add_last_error_fields_to_bgp_peers.php (hunk 1 above)
Comment (kedare) [reply database/migrations/2020_04_13_150500_add_last_error_fields_to_bgp_peers.php]: Done

**Comment by murrant** [comment]
Inserting additional data into existing snmprec files can be tricky.  If you do it by hand, make sure the oid is always increasing.

**Comment by kedare** [comment]
Maybe it would be more interesting to let someone that has a lab environment to add new test data ? I don't have access to non-production environment and I am not allowed to dump data from production for the tests (for the risk of leaking data)

**Comment by murrant** [comment]
It would be good to have at least one with test data.

You could manually add the oids, but that seems a little tedious.

**Comment by zombah** [comment]
@kedare @murrant There is exist repo with snmpsim test data, i didn't checked it myself, but maybe it contain something useful for cases like this - https://github.com/etingof/snmpsim-data/tree/master/data/network/router

**Comment by kedare** [comment]
I added some test data, let me know if this is enough

**Comment by murrant** [comment]
Thanks @kedare , nice work :)

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-63-release-changelog-april-2020/11828/1

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-63-high-cpu-on-arista-switches-with-bgp-peers-module-enabled/11947/4
